File: kotlin/utilities/src/main/kotlin/org/gristle/pdxpuzzles/utilities/objects/UnionFind.kt
package org.gristle.pdxpuzzles.utilities.objects

class UnionFind(val parent: MutableList<Int>, val size: MutableList<Int>) {
    companion object {
        fun new(n: Int) = UnionFind(MutableList(n) { it }, MutableList(n) { 1 })
    }

    fun find(x: Int): Int {
        if (parent[x] != x) {
            parent[x] = find(parent[x])
        }
        return parent[x]
    }

    fun union(x: Int, y: Int): Boolean {
        val x = find(x)
        val y = find(y)

        if (x == y) {
            return false
        }
        if (size[x] >= size[y]) {
            parent[y] = x
            size[x] += size[y]
        } else {
            parent[x] = y
            size[y] += size[x]
        }
        return true
    }

    fun update() {
        for (i in parent.indices) {
            find(i)
        }
    }
}

File: rust/utilities/src/structs/union_find.rs
pub struct UnionFind {
    pub parent: Vec<usize>,
    pub size: Vec<usize>,
}

impl UnionFind {
    pub fn new(n: usize) -> Self {
        Self {
            parent: (0..n).collect(),
            size: vec![1; n],
        }
    }

    pub fn find(&mut self, x: usize) -> usize {
        if self.parent[x] != x {
            self.parent[x] = self.find(self.parent[x]);
        }
        self.parent[x]
    }

    pub fn union(&mut self, x: usize, y: usize) -> bool {
        let x = self.find(x);
        let y = self.find(y);

        if x == y {
            return false;
        }
        if self.size[x] >= self.size[y] {
            self.parent[y] = x;
            self.size[x] += self.size[y];
        } else {
            self.parent[x] = y;
            self.size[y] += self.size[x];
        }
        true
    }

    pub fn update(&mut self) {
        for i in 0..self.parent.len() {
            self.find(i);
        }
    }
}